crates/ty_python_semantic/src/semantic_index/builder.rs

```diff
@@ -2654,3 +2654,76 @@ fn dunder_all_extend_argument(value: &ast::Expr) -> Option<&ast::Expr> {
 
     (attr == "__all__").then_some(value)
 }
+
+/// Builds an interval-map that matches expressions (by their node index) to their enclosing scopes.
+///
+/// The interval map is built in a two-step process because the expression ids are assigned in source order,
+/// but we visit the expressions in semantic order. Few expressions are registered out of order.
+///
+/// 1. build a point vector that maps node indices to their corresponding file scopes. The
+///    vector is sorted in ascending order of node indices and uses insertion sort to maintain the order.
+///    Insertion sort is efficient because most expressions are registered in order.
+/// 2. Condense the point vector into an interval map by collapsing adjacent node indices with the same scope
+///    into a single interval.
+struct ExpressionsScopeMapBuilder {
+    expression_and_scope: Vec<(NodeIndex, FileScopeId)>,
+}
+
+impl ExpressionsScopeMapBuilder {
+    fn new() -> Self {
+        Self {
+            expression_and_scope: vec![],
+        }
+    }
+
+    fn record_expression(&mut self, expression: &impl HasTrackedScope, scope: FileScopeId) {
+        let expression_index = expression.node_index().load();
+
+        if self
+            .expression_and_scope
+            .last()
+            .is_some_and(|last| last.0 <= expression_index)
+        {
+            self.expression_and_scope.push((expression_index, scope));
+            return;
+        }
+
+        let insertion_point = self
+            .expression_and_scope
+            .iter()
+            .rposition(|(index, _)| *index <= expression_index)
+            .map(|index| index + 1)
+            .unwrap_or_default();
```

Comment:
Using `rposition` may require walking large distances back if the inner interleaving scopes are large? Would it make sense to use a binary search here? Something like
```suggestion
        if self
            .expression_and_scope
            .last()
            .is_none_or(|last| last.0 <= expression_index)
        {
            self.expression_and_scope.push((expression_index, scope));
            return;
        }

        if self.expression_and_scope[0].0 > expression_index {
            self.expression_and_scope
                .insert(0, (expression_index, scope));
            return;
        }

        let insertion_point = self
            .expression_and_scope
            .partition_point(|(index, _)| *index <= expression_index);
```